clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -80,15 +156,34 @@ public Errors error() {
         return Errors.forCode(data.errorCode());
     }
 
-    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> responseData() {
+    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> responseData(Map<Uuid, String> topicNames, short version) {
+        if (version < 13)
+            return toResponseDataMap();
+        return toResponseDataMap(topicNames);
+
+    }
+
+    // TODO: Should be replaced or cleaned up. The idea is that in KafkaApis we need to reconstruct responseData even though we could have just passed in and out a map.
+    //  With topic IDs, recreating the map takes a little more time since we have to get the topic name from the topic ID to name map.
+    //  The refactor somewhat helps in KafkaApis, but we have to recompute the map instead of just returning it.
+    //  This is unsafe in test cases now (FetchSessionTest) where it used to be safe. Before it would just pull the responseData map.
+    //  If we wanted to recompute with topicNames we could call responseData(topicNames) however,
+    //  now it will just return the version computed here.
+
+    // Used when we can guarantee responseData is populated with all possible partitions
+    // This occurs when we have a response version < 13 or we built the FetchResponse with
+    // responseDataMap as a parameter and we have the same topic IDs available.
+    public LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> resolvedResponseData() {
```

Comment:
Yes. We may have topic IDs that could not be resolved. These topics will not have a topic name. I agree that this isn't the cleanest solution, but it was the one that worked while keeping TopicPartitions as keys for the map.

Reply:
Maybe we should remove the usage of this method first.

Is it useful if we make all callers use auto-generated data (topic and partition) instead of this map (TopicPartition)?

Reply:
I can remove it. I was keeping it as a placeholder for now. I was hoping to find a way to not regenerate the map that we just passed in to make the FetchResponse. 

I'm also not quite sure what you mean by "auto-generated data"

Reply:
> I'm also not quite sure what you mean by "auto-generated data"

Make all callers use FetchResponseData instead of Map[TopicPartition, ...]. In other words, FetchResponse does not return TopicPartition anymore.

Reply:
Ah I see. I'd have to take a look. I'm wondering if we would still need to convert to topic names for certain methods.

Reply:
Yeah, so for example, TopicPartitions are given to ReplicaManager.fetchMessages where it is used heavily. I think we still want the map for now. 

One thing I was confused about with this code is why we generate `unconvertedFetchResponse` with `updateAndGenerateResponseData`. We pass in the map to create `unconvertedFetchResponse`. With the optimization to lazily compute the map, recompute it a few lines of code later in `createResponse`. I'm hoping to avoid this somehow. (I think the optimization is great everywhere else though and I was heading in that direction originally with this PR)

Reply:
> One thing I was confused about with this code is why we generate unconvertedFetchResponse with updateAndGenerateResponseData. We pass in the map to create unconvertedFetchResponse. With the optimization to lazily compute the map, recompute it a few lines of code later in createResponse. I'm hoping to avoid this somehow. (I think the optimization is great everywhere else though and I was heading in that direction originally with this PR)

you are right. I feel it is the side-effect caused mixing `TopicPartition` and generated code in production. The callers have to create a heavy copy to update some fields of passed data. 

BTW, that is what https://github.com/apache/kafka/pull/10269 want to resolve. The PR makes all code use auto-generated data so they can update fields of passed auto-generated data instead of creating a copy.

Reply:
Got it. I think the issue here then is that some of the information can not be auto-generated data. We need topic names for certain methods but names will not be in the auto-generated data.

Reply:
I suppose we could set them server side, but when we iterate though the topics, we would need to not include those whose topic IDs could not be resolved.